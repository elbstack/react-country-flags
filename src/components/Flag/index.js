import React, { PureComponent as Component, createElement } from 'react';
import PropTypes from 'prop-types';
import flags from '../../flags/react';

const fetchFlag = async country => {
  if (!flags[country]) return Promise.resolve(null);
  const { default: data } = await flags[country]();
  return data;
};

// thanks to https://github.com/facebook/react/issues/5465#issuecomment-157888325
const makeCancelable = promise => {
  let hasCanceled_ = false;

  const wrappedPromise = new Promise((resolve, reject) => {
    promise.then(
      val => (hasCanceled_ ? reject({ isCanceled: true }) : resolve(val)),
      error => (hasCanceled_ ? reject({ isCanceled: true }) : reject(error))
    );
  });

  return {
    promise: wrappedPromise,
    cancel() {
      hasCanceled_ = true;
    },
  };
};

export const countries = (() => Object.keys(flags))();

export default class Flags extends Component {
  cancelUpdate = () => {};

  state = {
    flag: null,
  };

  static defaultProps = {
    country: null,
    asSquare: false,
  };

  componentDidMount() {
    this.updateFlag();
  }

  componentDidUpdate(prevProps) {
    const { country } = this.props;

    if (country !== prevProps.country) {
      this.updateFlag();
    }
  }

  componentWillUnmount() {
    this.cancelUpdate();
  }

  updateFlag() {
    const { country } = this.props;
    const { promise, cancel } = makeCancelable(fetchFlag(country));
    promise.then(flag => {
      this.setState({ flag });
    });
    this.cancelUpdate = cancel;
  }

  render() {
    const { asSquare, country, ...props } = this.props;
    const { flag } = this.state;
    if (!flag && !asSquare) return null;
    if (!asSquare) return createElement(flag, props);
    return (
      <svg viewBox="0 0 100 100" preserveAspectRatio="xMidYMid" {...props}>
        {!flag ? null : createElement(flag)}
      </svg>
    );
  }
}

Flags.propTypes = {
  country: PropTypes.string,
  asSquare: PropTypes.bool,
};
