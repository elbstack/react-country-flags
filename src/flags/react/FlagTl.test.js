import React from 'react';
import renderer from 'react-test-renderer';
import Flag from './FlagTl';

it('FlagTl renders without crashing', () => {
  const component = renderer.create(<Flag />);

  let tree = component.toJSON();
  expect(tree).toMatchSnapshot();
});
