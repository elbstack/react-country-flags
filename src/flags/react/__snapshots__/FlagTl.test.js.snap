// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`FlagTl renders without crashing 1`] = `
<svg
  viewBox="-3 -6 24 12"
  xmlnsXlink="http://www.w3.org/1999/xlink"
>
  <path
    d="M21 6V-6H-3V6z"
    fill="#dc241f"
  />
  <path
    d="M-3-6V6L9 0z"
    fill="#ffc726"
  />
  <path
    d="M-3-6V6l8-6z"
  />
  <g
    transform="rotate(-26.565)"
  >
    <g
      id="s21cd84c8"
    >
      <g
        id="s74890357"
      >
        <path
          d="M0-2.1V0h1z"
          fill="#fff"
          id="s152cec9a"
          transform="rotate(18 0 -2.1)"
        />
        <use
          transform="scale(-1 1)"
          xlinkHref="#s152cec9a"
        />
      </g>
      <use
        transform="rotate(72)"
        xlinkHref="#s74890357"
      />
    </g>
    <use
      transform="rotate(-72)"
      xlinkHref="#s74890357"
    />
    <use
      transform="rotate(144)"
      xlinkHref="#s21cd84c8"
    />
  </g>
</svg>
`;
